tachyon/math/circle/circle_point.h

```diff
@@ -0,0 +1,130 @@
+      : CirclePoint(std::move(point.x), std::move(point.y)) {}
+  constexpr CirclePoint(const BaseField& x, const BaseField& y)
+      : x_(x), y_(y) {}
+  constexpr CirclePoint(BaseField&& x, BaseField&& y)
+      : x_(std::move(x)), y_(std::move(y)) {}
+
+  constexpr static CirclePoint CreateChecked(const BaseField& x,
+                                             const BaseField& y) {
+    CirclePoint ret = {x, y};
+    CHECK(ret.IsOnCircle());
+    return ret;
+  }
+
+  constexpr static CirclePoint CreateChecked(BaseField&& x, BaseField&& y) {
+    CirclePoint ret = {std::move(x), std::move(y)};
+    CHECK(ret.IsOnCircle());
+    return ret;
+  }
+
+  constexpr static CirclePoint Zero() { return CirclePoint(); }
+
+  constexpr static CirclePoint Generator() {
+    return {Circle::Config::kGenerator.x, Circle::Config::kGenerator.y};
+  }
+
+  constexpr static CirclePoint Random() {
+    return ScalarField::Random() * Generator();
+  }
+
+  constexpr const BaseField& x() const { return x_; }
+  constexpr const BaseField& y() const { return y_; }
+
+  constexpr bool operator==(const CirclePoint& other) const {
+    return x_ == other.x_ && y_ == other.y_;
+  }
+  constexpr bool operator!=(const CirclePoint& other) const {
+    return !operator==(other);
+  }
+
+  constexpr bool IsZero() const { return x_.IsOne() && y_.IsZero(); }
+
+  constexpr bool IsOnCircle() { return Circle::IsOnCircle(*this); }
+
+  constexpr CirclePoint Conjugate() const { return {x_, -y_}; }
+
+  constexpr CirclePoint ComplexConjugate() const {
+    return {x_.ComplexConjugate(), y_.ComplexConjugate()};
+  }
+
+  constexpr CirclePoint Antipode() const { return {-x_, -y_}; }
+
+  std::string ToString() const {
+    return absl::Substitute("($0, $1)", x_.ToString(), y_.ToString());
+  }
+
+  std::string ToHexString(bool pad_zero = false) const {
+    return absl::Substitute("($0, $1)", x_.ToHexString(pad_zero),
+                            y_.ToHexString(pad_zero));
+  }
+
+  // AdditiveSemigroup methods
+  constexpr CirclePoint Add(const CirclePoint& other) const {
+    return {x_ * other.x_ - y_ * other.y_, x_ * other.y_ + y_ * other.x_};
+  }
+
+  constexpr CirclePoint& AddInPlace(const CirclePoint& other) {
+    return *this = {x_ * other.x_ - y_ * other.y_,
+                    x_ * other.y_ + y_ * other.x_};
+  }
+
+  // AdditiveGroup methods
+  constexpr CirclePoint Sub(const CirclePoint& other) const {
+    return {x_ * other.x_ + y_ * other.y_, -x_ * other.y_ + y_ * other.x_};
+  }
+
+  constexpr CirclePoint& SubInPlace(const CirclePoint& other) {
+    return *this = {x_ * other.x_ + y_ * other.y_,
+                    -x_ * other.y_ + y_ * other.x_};
+  }
+
+  constexpr CirclePoint Negative() const { return {x_, -y_}; }
+
+  constexpr CirclePoint& NegInPlace() {
+    y_.NegInPlace();
+    return *this;
+  }
+
+  constexpr CirclePoint operator*(const ScalarField& idx) const {
+    return this->ScalarMul(idx.value());
```

Comment:
Is there any reason to name the parameter as `idx`? What about `scalar`?

Reply:
I think this is because stwo name `ScalarField` as `CirclePointIndex`, but yours is more reasonable.